[QNN EP] Fix Pad UT

winskuo-quic opened the pull request:
### Description

QNN EP has 2 unit tests failing:

TEST_F(QnnHTPBackendTests, DISABLED_PadReflectMode)
TEST_F(QnnHTPBackendTests, DISABLED_Pad4dOutOfRangePadConstantValue)

For the first unit test, in QNN's master definition, it is stated that when using MIRROR_REFLECT, the before and after pad amounts must not be greater than shape(in[0])[i] - 1. Therefore, we need to change the pad amount from {0,2,0,0} to {0,1,0,0}.

For second unit test, QNN does not have limitations stating that pad constant should be smaller than input[0]. The reason that the test is failing is because the unit test did not take the pad constant into consideration when doing quantization.

### Motivation and Context
Fix the 2 unit tests mentioned in description.

HectorSVC commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-python-checks-ci-pipeline,onnxruntime-binary-size-checks-ci-pipeline

HectorSVC commented:
1. For the 1st test, if that's the case, we need to do some check in PadOpBuilder to detect this case and fall back to CPU EP.
2. For the 2nd test, the constant_value is quantized because we set bool constant_value_quantized = true by default, refers to code line 124. And in PadOpBuilder we always use de-quantize it and use float32 for constant_value in QNN.

chiwwang commented:
cc myself

winskuo-quic commented:
Hi Hector,

Thanks for the suggestions.
Let me elaborate a little bit more on the 2 issues:

1. For 1st test, indeed we should be checking for this case in PadOpBuilder. I will work on it and try coming up with a solution or another commit.
2. For 2nd test, please refer to the QNN documentation under HTP Operation section, which specifies the constraints for pad_constant_value. pad_constant_value must be quantizable with quantization params same as input. For this test case, since pad_constant_value cannot be quantized correctly with the input's scale&offset, the UT will have undesired output.

HectorSVC commented:
Regarding pad_constant_value, my understanding is if it's quantized data, we should store the data using QNN_DATATYPE_INT_32. From another hand, it also support non-quantized data using QNN_DATATYPE_FLOAT_32. In the PadOpBuilder, we always set it to non-quantized data using QNN_DATATYPE_FLOAT_32. I found Qnn-onnx-converter does the same thing.

---
In reply to: [1767766209](https://github.com/microsoft/onnxruntime/pull/17982#issuecomment-1767766209) [](ancestors = 1767766209)

winskuo-quic commented:
@microsoft-github-policy-service agree company="Qualcomm"

winskuo-quic commented:
I will focus more on the limitations rather than changing the value to make UT pass in the future.

Updates on the 2 HTP unit tests:
TEST_F(QnnHTPBackendTests, DISABLED_PadReflectMode)
- I have added a loop that checks if before and after pad amounts are greater than shape(in[0])[i] - 1.
- I have added a couple more test cases to ensure behavior is desired.

TEST_F(QnnHTPBackendTests, DISABLED_Pad4dOutOfRangePadConstantValue)
- I have changed the constant value from 8 back to 9.

HectorSVC commented:
>                            ExpectedEPNodeAssignment::All,

you can set it to ExpectedEPNodeAssignment::None which means the nodes not assigned to QNN EP.
So that you can remove the "DISABLED_".

---
Refers to: onnxruntime/test/providers/qnn/pad_op_test.cpp:286 in 4aae07e. [](commit_id = 4aae07e0788f4a4cdfe72a9f32e97d8cc7541f75, deletion_comment = False)

adrianlizarraga commented on onnxruntime/test/providers/qnn/pad_op_test.cpp:
```diff
@@ -167,7 +167,7 @@ TEST_F(QnnCPUBackendTests, Pad2dPadsNotIni) {
 TEST_F(QnnCPUBackendTests, DISABLED_PadModeReflect) {
```
Should this test be enabled now?

winskuo-quic replied:
This test is still failing and should be DISABLED.
The CPU team has identified the root cause, and they will either fix this issue or update the documentation.

adrianlizarraga commented on onnxruntime/test/providers/qnn/pad_op_test.cpp:
```diff
@@ -266,8 +266,18 @@ TEST_F(QnnHTPBackendTests, PadHasConstantValueQuantized) {
+}
+
+// Pad amount should not be greater than shape(input[0])[i] - 1
+TEST_F(QnnHTPBackendTests, DISABLED_PadReflectModeOutOfRangePadAmount) {
```
Instead of disabling this test, we can pass `ExpectedEPNodeAssignment::None` to `RunQDQPadOpTest` to test that QNN EP rejects invalid padding.

winskuo-quic replied:
Hi @adrianlizarraga ,
Thanks for the feedback.
I have pushed a new commit which enabled the test and set ExpectedEPNodeAssignment to None.

adrianlizarraga reviewed (commented):
@winskuo-quic Thank you for contributing! Two minor comments.

adrianlizarraga commented:
/azp run Windows ARM64 QNN CI Pipeline,Windows x64 QNN CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline

adrianlizarraga commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-python-checks-ci-pipeline,onnxruntime-binary-size-checks-ci-pipeline

HectorSVC reviewed (approved):
:shipit:
